upload files to new branch for PR

janessatran opened the pull request.

slabounty commented:
Morning @janessatran  ... let's start here by reformatting your code to two spaces from four. That's more or less the ruby standard and will change things so we can take a look at some other suggestions.

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
+  lowercase_letters = Array('a'..'z')
+  uppercase_letters = Array('A'..'Z')
+  new_str = ''
+  str.each_char do |c|
```
Take a look at `each_with_object` for this piece. It's the way to go when you're building up something (a string, a hash, counting things, etc.) from the elements in an array.

janessatran replied:
When I try using each_with_object for the string, I get an error: 

```ruby
   undefined method `each_with_object' for "abc":String
```

Do I have to convert the string into an array first? And then do something like 
```ruby
  str_arry.each_with_object("") do |c, new_str|
     if condition
      new_str += ... 
```

slabounty replied:
That would work. You can check out `split` to create an array of characters.

slabounty replied:
So ... `str.split("").each_with_object("")  ... `

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
+  new_str = ''
+  str.each_char do |c|
+    if lowercase_letters.include?(c)
+      shifted_idx = lowercase_letters.index(c) + shift
```
We can see that both the `then` and the `else` sections of this are doing pretty much the same thing. Is there a way to pull them into another method maybe?

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
+    if lowercase_letters.include?(c)
+      shifted_idx = lowercase_letters.index(c) + shift
+      # if the new idx is out of bounds, loop back to beginning for idx
+      if shifted_idx > 26
```
So ... we use `26` a fair number of times in here. We can make it a constant I think. Or, even better, it's *really* representing the size of the array you're going through. Maybe use `lowercase.size` and `uppercase.size`?

slabounty commented on spec/caesar_cipher_spec.rb:
```diff
@@ -1,19 +1,12 @@
-describe "#caesar_cipher" do
+  it "applies Caesar Cipher encoding to your string" do
+    expect(caesar_cipher("What a string!", 5)).to eq("Bmfy f xywnsl!")
+  end
```
Generally, we like to see a blank line between the tests. Here and below.

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
-        else
-            new_str += c       
-        end
+  lowercase_letters = Array('a'..'z')
```
More "rubyesque" might be ...
```
lowercase_letters = ("a".."z").to_a
```

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,28 @@
+  new_str = ''
+  str.each_char do |c|
+    if $lowercase_letters.include?(c)
+      new_str += $lowercase_letters[get_shifted_idx('lowercase', c, shift)]
```
How about rather than passing in `'lowercase'` or `'uppercase'`, you just passed the array directly? I think that might allow you to get rid of your `if/else` in the `get_shifted_idx` method?

janessatran replied:
wow!! agreed! thanks for the suggestion

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,28 @@
+$lowercase_letters = ('a'..'z').to_a
```
I think you can just leave these in the `ceasar_cipher` method if you pass them to the `get_shifted_idx` method.

janessatran replied:
yup yup yup, thank you!

janessatran commented:
I have a question about how to decide between "readable" code and trying to confine things down into less lines of code. Which should I aim for typically? 

For example, I can do the following which has more lines of code, but to me is more readable (maybe to more experienced programmers it's different?)
```
  str.split("").each_with_object("") do |c| 
    if lowercase_letters.include?(c)
      cc_str << lowercase_letters[get_shifted_idx(lowercase_letters, c, shift)]
    elsif uppercase_letters.include?(c)
      cc_str << uppercase_letters[get_shifted_idx(uppercase_letters, c, shift)]
    else
      cc_str << c
    end
  end
```

or this, which is less "lines" of code but to me is less readable: 
```
cc_str = str.split("").each_with_object([]) { |c, new_str| new_str << (lowercase_letters.include?(c) ? lowercase_letters[get_shifted_idx(lowercase_letters, c, shift)] : uppercase_letters.include?(c) ? uppercase_letters[get_shifted_idx(uppercase_letters, c, shift)] : c) } 
```

slabounty commented:
I really like that you're thinking about the readability question. If you're in doubt, at all, go for readability. In the example above, the fact that you have a nested ternary operator is probably signal enough you've gone in the wrong direction. So ... the way you did it is definitely the way to go.

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
-        else
-            new_str += c       
-        end
+  cc_str = ''
```
I think you want to remove this.

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
+  cc_str = ''
+  lowercase_letters = ('a'..'z').to_a
+  uppercase_letters = ('A'..'Z').to_a
+  str.split("").each_with_object("") do |c| 
```
```
str.split("").each_with_object("") do |c, cc_str|
```

here.

slabounty commented on caesar_cipher.rb:
```diff
@@ -1,25 +1,24 @@
     end
-    return new_str
+  end
+  return cc_str
```
And I think you can remove this also. The result of `str.split("").each_with_object("") do |c, cc_str|` will be the `cc_str`. And, you have good tests, so you can make sure I'm not lying to you.

janessatran replied:
😱 magic!

janessatran commented:
Okay, got it! Thank you! Readability, making functions do one thing, learning lots :)
